### PR TITLE
Disallow indexed reference types in events when using ABIEncoderV2

### LINKED ISSUE #4687: Re-check structs in events with new encoder
- [x] check indexed
 - [x] check non-indexed

I think indexed should be disallowed even with new encoder

### PR DESCRIPTION
**Pull request by leonardoalt**
Closes #4687

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -895,7 +895,17 @@ bool TypeChecker::visit(EventDefinition const& _eventDef)
 	for (ASTPointer<VariableDeclaration> const& var: _eventDef.parameters())
 	{
 		if (var->isIndexed())
+		{
 			numIndexed++;
+			if (
+				_eventDef.sourceUnit().annotation().experimentalFeatures.count(ExperimentalFeature::ABIEncoderV2)
+				&& dynamic_cast<ReferenceType const*>(type(*var).get())
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: `&&` to the previous line

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -895,7 +895,17 @@ bool TypeChecker::visit(EventDefinition const& _eventDef)
...
+			)
+				m_errorReporter.typeError(
+					var->location(),
+					"Reference types cannot be indexed."
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: `Indexed reference types cannot yet be used with ABIEncoderV2`

**Review by chriseth — CHANGES_REQUESTED** [review]
Please add an endtoend test for dynamically-sized arrays both indexed and non-indexed and for both encoders (if supported) and a non-indexed struct with the new encoder.

**Comment by chriseth** [comment]
Oh and the struct should be taken from storage.

**Comment by chriseth** [comment]
Ah and I think we need an entry in the bug list don't we?

**Comment by leonardoalt** [comment]
@chriseth Let's merge this without the json path test and I'll create a separate PR for it.

**Review by chriseth — APPROVED** [review]

**Comment by axic** [comment]
I think this was working for strings, but not for structs.

Because the actual hashing is in done in `ExpressionCompiler` prior to abi encoding and that should work for strings. Well, it does use the specific "packed encoding" on it. Wonder if that is desired.

As it turned out in #5102 some tests rely on it.

**Pull request merged**